### PR TITLE
Auth joining

### PR DESCRIPTION
**Pull request by olsona**
Made new OAuth2{Config, DataGenerator, ServiceExtension} classes.  So far, Google and Instagram work.  

The plan is to combine all the service-level auth flows where possible to allow for more uniform and easy-to-understand auth flows based on underlying protocol, and to reduce reliance on third-party SDKs for future secure deployment environments (see #553).  Once all OAuth2 services are combined and verified, I will remove the old auth code; until then, there will be two auth flows coexisting.  When OAuth2 is done, I will do the same with OAuth1 services.

### CODE REVIEW

File: extensions/auth/portability-auth-facebook/src/main/java/org/datatransferproject/auth/facebook/FacebookOAuthConfig.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+package org.datatransferproject.auth.facebook;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.util.List;
+import java.util.Map;
+import org.datatransferproject.auth.oauth2.OAuth2Config;
+
+public class FacebookOAuthConfig implements OAuth2Config {
```
Comment (holachuy) [inline extensions/auth/portability-auth-facebook/src/main/java/org/datatransferproject/auth/facebook/FacebookOAuthConfig.java]: Let's add a comment with link to the facebook api documentation site.

File: extensions/auth/portability-auth-facebook/src/main/java/org/datatransferproject/auth/facebook/FacebookOAuthConfig.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+  @Override
+  public String getServiceName() {
+    return "Facebook";
```
Comment (holachuy) [inline extensions/auth/portability-auth-facebook/src/main/java/org/datatransferproject/auth/facebook/FacebookOAuthConfig.java]: Is this for display or a canonical value that should be all in the same case?

File: extensions/auth/portability-auth-google/src/main/java/org/datatransferproject/auth/google/GoogleOAuthConfig.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+package org.datatransferproject.auth.google;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.util.List;
+import java.util.Map;
+import org.datatransferproject.auth.oauth2.OAuth2Config;
+
+public class GoogleOAuthConfig implements OAuth2Config {
```
Comment (holachuy) [inline extensions/auth/portability-auth-google/src/main/java/org/datatransferproject/auth/google/GoogleOAuthConfig.java]: Same comments here on documentation and values below

File: extensions/auth/portability-auth-instagram/src/main/java/org/datatransferproject/auth/instagram/InstagramOAuthConfig.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+package org.datatransferproject.auth.instagram;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.datatransferproject.auth.oauth2.OAuth2Config;
+
+public class InstagramOAuthConfig implements OAuth2Config {
```
Comment (holachuy) [inline extensions/auth/portability-auth-instagram/src/main/java/org/datatransferproject/auth/instagram/InstagramOAuthConfig.java]: Same comment here

File: extensions/auth/portability-auth-instagram/src/main/java/org/datatransferproject/auth/instagram/InstagramOAuthConfig.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  }
+
+  @Override
+  public Map<String, List<String>> getExportScopes() {
+    return ImmutableMap.of("PHOTOS", ImmutableList.of("basic"));
+  }
+
+  @Override
+  public Map<String, List<String>> getImportScopes() {
+    return ImmutableMap.of();
```
Comment (holachuy) [inline extensions/auth/portability-auth-instagram/src/main/java/org/datatransferproject/auth/instagram/InstagramOAuthConfig.java]: Comment here

File: extensions/auth/portability-auth-microsoft/src/main/java/org/datatransferproject/auth/microsoft/MicrosoftOAuthConfig.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+package org.datatransferproject.auth.microsoft;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.util.List;
+import java.util.Map;
+import org.datatransferproject.auth.oauth2.OAuth2Config;
+
+public class MicrosoftOAuthConfig implements OAuth2Config {
```
Comment (holachuy) [inline extensions/auth/portability-auth-microsoft/src/main/java/org/datatransferproject/auth/microsoft/MicrosoftOAuthConfig.java]: Same comment

File: extensions/auth/portability-auth-oauth2/src/main/java/org/datatransferproject/auth/oauth2/OAuth2Config.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+ * limitations under the License.
+ */
+
+package org.datatransferproject.auth.oauth2;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public interface OAuth2Config {
```
Comment (holachuy) [inline extensions/auth/portability-auth-oauth2/src/main/java/org/datatransferproject/auth/oauth2/OAuth2Config.java]: Document this interface

File: extensions/auth/portability-auth-oauth2/src/main/java/org/datatransferproject/auth/oauth2/OAuth2Config.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+  String getServiceName();
+
+  String getAuthUrl();
+
+  String getTokenUrl();
+
+  Map<String, List<String>> getExportScopes();
```
Comment (holachuy) [inline extensions/auth/portability-auth-oauth2/src/main/java/org/datatransferproject/auth/oauth2/OAuth2Config.java]: Does ordering or duplicates matter? Should it be a Set or OrderedSet instead?

Comment (holachuy) [reply extensions/auth/portability-auth-oauth2/src/main/java/org/datatransferproject/auth/oauth2/OAuth2Config.java]: Add comment including the first value is 'transferType' as defined in the auth data generators and elsewhere.

File: extensions/auth/portability-auth-twitter/src/main/java/org/datatransferproject/auth/twitter/TwitterOAuthConfig.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+package org.datatransferproject.auth.twitter;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import java.util.List;
+import java.util.Map;
+import org.datatransferproject.auth.oauth2.OAuth2Config;
+
+public class TwitterOAuthConfig implements OAuth2Config {
```
Comment (holachuy) [inline extensions/auth/portability-auth-twitter/src/main/java/org/datatransferproject/auth/twitter/TwitterOAuthConfig.java]: Same

File: extensions/auth/portability-auth-facebook/src/main/java/org/datatransferproject/auth/facebook/FacebookOAuthConfig.java (hunk 2 above)
Comment (olsona) [reply extensions/auth/portability-auth-facebook/src/main/java/org/datatransferproject/auth/facebook/FacebookOAuthConfig.java]: I believe it's functionally only for display, aside from the client ids and secrets (which are the service name in all-caps)

File: extensions/auth/portability-auth-google/src/main/java/org/datatransferproject/auth/google/GoogleOAuthConfig.java (hunk 3 above)
Comment (olsona) [reply extensions/auth/portability-auth-google/src/main/java/org/datatransferproject/auth/google/GoogleOAuthConfig.java]: Done.

File: extensions/auth/portability-auth-instagram/src/main/java/org/datatransferproject/auth/instagram/InstagramOAuthConfig.java (hunk 4 above)
Comment (olsona) [reply extensions/auth/portability-auth-instagram/src/main/java/org/datatransferproject/auth/instagram/InstagramOAuthConfig.java]: Done.

File: extensions/auth/portability-auth-instagram/src/main/java/org/datatransferproject/auth/instagram/InstagramOAuthConfig.java (hunk 5 above)
Comment (olsona) [reply extensions/auth/portability-auth-instagram/src/main/java/org/datatransferproject/auth/instagram/InstagramOAuthConfig.java]: Done.

File: extensions/auth/portability-auth-oauth2/src/main/java/org/datatransferproject/auth/oauth2/OAuth2Config.java (hunk 8 above)
Comment (olsona) [reply extensions/auth/portability-auth-oauth2/src/main/java/org/datatransferproject/auth/oauth2/OAuth2Config.java]: Order does not seem to matter, but I can't imagine duplicates would be a good thing.

File: extensions/auth/portability-auth-google/src/main/java/org/datatransferproject/auth/google/GoogleOAuthConfig.java (hunk 3 above)
Comment (olsona) [reply extensions/auth/portability-auth-google/src/main/java/org/datatransferproject/auth/google/GoogleOAuthConfig.java]: Sorry, I only added documentation to the methods, not the class.

File: libraries/auth/src/main/java/org/datatransferproject/auth/OAuth2ServiceExtension.java [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+import org.datatransferproject.api.launcher.ExtensionContext;
+import org.datatransferproject.spi.api.auth.AuthDataGenerator;
+import org.datatransferproject.spi.api.auth.AuthServiceProviderRegistry.AuthMode;
+import org.datatransferproject.spi.api.auth.extension.AuthServiceExtension;
+import org.datatransferproject.spi.cloud.storage.AppCredentialStore;
+import org.datatransferproject.types.transfer.auth.AppCredentials;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class OAuth2ServiceExtension implements AuthServiceExtension {
```
[the remark this answers is not in the document]
Comment (olsona) [reply libraries/auth/src/main/java/org/datatransferproject/auth/OAuth2ServiceExtension.java]: Done.

File: libraries/auth/src/main/java/org/datatransferproject/auth/OAuth2DataGenerator.java [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+import java.io.IOException;
+import java.util.Set;
+import org.datatransferproject.spi.api.auth.AuthDataGenerator;
+import org.datatransferproject.spi.api.auth.AuthServiceProviderRegistry.AuthMode;
+import org.datatransferproject.spi.api.types.AuthFlowConfiguration;
+import org.datatransferproject.types.transfer.auth.AppCredentials;
+import org.datatransferproject.types.transfer.auth.AuthData;
+import org.datatransferproject.types.transfer.auth.TokensAndUrlAuthData;
+
+public class OAuth2DataGenerator implements AuthDataGenerator {
```
[the remark this answers is not in the document]
Comment (olsona) [reply libraries/auth/src/main/java/org/datatransferproject/auth/OAuth2DataGenerator.java]: Done.

File: extensions/auth/portability-auth-google/src/main/java/org/datatransferproject/auth/google/GoogleOAuthConfig.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+  // https://developers.google.com/identity/protocols/OAuth2WebServer
+
+  @Override
+  public String getServiceName() {
+    return "Google";
+  }
+
+  @Override
+  // See https://developers.google.com/identity/protocols/OAuth2WebServer#creatingclient
```
[the remark this answers is not in the document]
Comment (olsona) [reply extensions/auth/portability-auth-google/src/main/java/org/datatransferproject/auth/google/GoogleOAuthConfig.java]: Done.

File: libraries/auth/src/main/java/org/datatransferproject/auth/OAuth2ServiceExtension.java [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+  private volatile Map<String, OAuth2DataGenerator> exportAuthDataGenerators;
+  private volatile Map<String, OAuth2DataGenerator> importAuthDataGenerators;
+
+  private AppCredentials appCredentials;
+  private HttpTransport httpTransport;
+
+  private boolean initialized = false;
+
+  public OAuth2ServiceExtension(OAuth2Config oAuth2Config) {
+    this.oAuth2Config = oAuth2Config;
```
Comment (holachuy) [inline libraries/auth/src/main/java/org/datatransferproject/auth/OAuth2ServiceExtension.java]: We should probably validate that all the data is present in the config before continuing construction.

**Pull request merged**